resources/type-system/inference.md

```diff
@@ -299,61 +308,137 @@ The function **futureValueTypeSchema** is defined as follows:
-inferred type of the expression body, using the local type inference algorithm
-described below with a typing context as computed above.
-
-The actual returned type of a function literal with a block body is computed as
-follows.  Let `T` be `Never` if every control path through the block exits the
-block without reaching the end of the block, as computed by the **definite
-completion** analysis specified elsewhere.  Let `T` be `Null` if any control
-path reaches the end of the block without exiting the block, as computed by the
-**definite completion** analysis specified elsewhere.  Let `K` be the typing
-context for the function body as computed above from the imposed return type
-schema.
-  - For each `return e;` statement in the block, let `S` be the inferred type of
-    `e`, using the local type inference algorithm described below with typing
-    context `K`, and update `T` to be `UP(flatten(S), T)` if the enclosing
-    function is `async`, or `UP(S, T)` otherwise.
-  - For each `return;` statement in the block, update `T` to be `UP(Null, T)`.
-  - For each `yield e;` statement in the block, let `S` be the inferred type of
-    `e`, using the local type inference algorithm described below with typing
-    context `K`, and update `T` to be `UP(S, T)`.
-  - If the enclosing function is marked `sync*`, then for each `yield* e;`
-    statement in the block, let `S` be the inferred type of `e`, using the
-    local type inference algorithm described below with a typing context of
-    `Iterable<K>`; let `E` be the type such that `Iterable<E>` is a
-    super-interface of `S`; and update `T` to be `UP(E, T)`.
-  - If the enclosing function is marked `async*`, then for each `yield* e;`
-    statement in the block, let `S` be the inferred type of `e`, using the
-    local type inference algorithm described below with a typing context of
-    `Stream<K>`; let `E` be the type such that `Stream<E>` is a super-interface
-    of `S`; and update `T` to be `UP(E, T)`.
-
-The **actual returned type** of the function literal is the value of `T` after
-all `return` and `yield` statements in the block body have been considered.
-
-Let `T` be the **actual returned type** of a function literal as computed above.
+**actual value type** of the function literal. _The actual value type
+represents the types of the actual values returned by non-generator functions,
+and the values emitted by generator functions._
+
+Let `K` be the value context schema for the function body as computed above
+from the imposed return type schema.
+_When we refer to the inferred type of an expression with a typing context,
+it is the static type of the expression inferred using the local type inference
+algorithm described below._
+
+The actual value type of a function literal with an expression body, `=> e`,
+_(which cannot be a generator function)_ is computed as follows:
+  - If the enclosing function is marked `async`,
```

Comment:
I think "enclosing function" is ambiguous here - I think you mean "function which encloses `e`", but the text also supports a reading as "function which encloses the function literal".